# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/TexUnpackBlob.h
# Commit: ccfa767dba64
# Full Hash: ccfa767dba644dc193e0246eb8e8ff3377e8b8a5
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Client-side bindings mirror for precise CC, and merge similar codepaths. r=handyman
#   
#   * Context loss using RAII
#   * Move Program reflection Client-side
#   
# ==============================================================================

diff -r 71c122ac0ca7 -r ccfa767dba64 dom/canvas/TexUnpackBlob.h
--- a/dom/canvas/TexUnpackBlob.h	Wed Jan 08 22:19:14 2020 +0000
+++ b/dom/canvas/TexUnpackBlob.h	Wed Jan 08 22:19:16 2020 +0000
@@ -13,7 +13,6 @@
 
 namespace mozilla {
 
-class ClientWebGLContext;
 class UniqueBuffer;
 class WebGLContext;
 class WebGLTexture;
@@ -24,14 +23,6 @@
 class HTMLVideoElement;
 }  // namespace dom
 
-namespace ipc {
-template <typename T>
-struct PcqParamTraits;
-class ConsumerView;
-class ProducerView;
-struct PcqStatus;
-}  // namespace ipc
-
 namespace gfx {
 class DataSourceSurface;
 }  // namespace gfx
@@ -48,28 +39,29 @@
 
 class TexUnpackBlob {
  public:
-  uint32_t mAlignment = 0;
-  uint32_t mRowLength = 0;
-  uint32_t mImageHeight = 0;
-  uint32_t mSkipPixels = 0;
-  uint32_t mSkipRows = 0;
-  uint32_t mSkipImages = 0;
-  uint32_t mWidth = 0;
-  uint32_t mHeight = 0;
-  uint32_t mDepth = 0;
+  const uint32_t mAlignment;
+  const uint32_t mRowLength;
+  const uint32_t mImageHeight;
+  const uint32_t mSkipPixels;
+  const uint32_t mSkipRows;
+  const uint32_t mSkipImages;
+  const uint32_t mWidth;
+  const uint32_t mHeight;
+  const uint32_t mDepth;
 
-  gfxAlphaType mSrcAlphaType;
+  const gfxAlphaType mSrcAlphaType;
 
   bool mNeedsExactUpload;
 
  protected:
-  TexUnpackBlob(const WebGLPixelStore& pixelStore, TexImageTarget target,
+  TexUnpackBlob(const WebGLContext* webgl, TexImageTarget target,
                 uint32_t rowLength, uint32_t width, uint32_t height,
                 uint32_t depth, gfxAlphaType srcAlphaType);
 
-  // For IPC
-  TexUnpackBlob() {}
+ public:
+  virtual ~TexUnpackBlob() {}
 
+ protected:
   bool ConvertIfNeeded(WebGLContext* webgl, const uint32_t rowLength,
                        const uint32_t rowCount, WebGLTexelFormat srcFormat,
                        const uint8_t* const srcBegin, const ptrdiff_t srcStride,
@@ -79,10 +71,6 @@
                        UniqueBuffer* const out_anchoredBuffer) const;
 
  public:
-  virtual ~TexUnpackBlob() {}
-
-  virtual TexUnpackBytes* AsTexUnpackBytes() { return nullptr; }
-
   virtual bool HasData() const { return true; }
 
   virtual bool Validate(WebGLContext* webgl, const webgl::PackingInfo& pi) = 0;
@@ -100,18 +88,15 @@
 
 class TexUnpackBytes final : public TexUnpackBlob {
  public:
-  RawBuffer<const uint8_t> mPtr;
+  const bool mIsClientData;
+  const uint8_t* const mPtr;
+  const size_t mAvailBytes;
 
-  TexUnpackBytes(const WebGLPixelStore& pixelStore, TexImageTarget target,
+  TexUnpackBytes(const WebGLContext* webgl, TexImageTarget target,
                  uint32_t width, uint32_t height, uint32_t depth,
                  bool isClientData, const uint8_t* ptr, size_t availBytes);
 
-  // For IPC
-  TexUnpackBytes() {}
-
-  TexUnpackBytes* AsTexUnpackBytes() override { return this; }
-
-  virtual bool HasData() const override { return mPtr && mPtr.Data(); }
+  virtual bool HasData() const override { return !mIsClientData || bool(mPtr); }
 
   virtual bool Validate(WebGLContext* webgl,
                         const webgl::PackingInfo& pi) override;
@@ -125,12 +110,11 @@
 
 class TexUnpackImage final : public TexUnpackBlob {
  public:
-  RefPtr<layers::Image> mImage;
+  const RefPtr<layers::Image> mImage;
 
   TexUnpackImage(const WebGLContext* webgl, TexImageTarget target,
-                 uint32_t rowLength, uint32_t width, uint32_t height,
-                 uint32_t depth, layers::Image* image,
-                 gfxAlphaType srcAlphaType);
+                 uint32_t width, uint32_t height, uint32_t depth,
+                 layers::Image* image, gfxAlphaType srcAlphaType);
 
   ~TexUnpackImage();  // Prevent needing to define layers::Image in the header.
 
@@ -144,24 +128,14 @@
                              GLenum* const out_error) const override;
 };
 
-// If constructed from a surface, the surface is mapped as long as this object
-// exists.
 class TexUnpackSurface final : public TexUnpackBlob {
  public:
-  gfx::IntSize mSize;
-  gfx::SurfaceFormat mFormat;
-  RawBuffer<> mData;
-  int32_t mStride;
-  // Map may be null in host process because memory is mapped in the client
-  UniquePtr<gfx::DataSourceSurface::ScopedMap> mMap;
+  const RefPtr<gfx::DataSourceSurface> mSurf;
 
-  TexUnpackSurface(const WebGLPixelStore& pixelStore, TexImageTarget target,
+  TexUnpackSurface(const WebGLContext* webgl, TexImageTarget target,
                    uint32_t width, uint32_t height, uint32_t depth,
                    gfx::DataSourceSurface* surf, gfxAlphaType srcAlphaType);
 
-  // For PcqParamTraits
-  TexUnpackSurface() : mStride(0), mMap(nullptr) {}
-
   virtual bool Validate(WebGLContext* webgl,
                         const webgl::PackingInfo& pi) override;
   virtual bool TexOrSubImage(bool isSubImage, bool needsRespec,
@@ -173,7 +147,6 @@
 };
 
 }  // namespace webgl
-
 }  // namespace mozilla
 
 #endif  // TEX_UNPACK_BLOB_H_